debuginfod/src/client.rs

```diff
@@ -0,0 +1,136 @@
+        if res.status().is_success() {
+            res.copy_to(dest)?;
+            Ok(dest.flush()?)
+        } else {
+            Err(anyhow!(format!("request error {}", res.status())))
+        }
+    }
+
+    /// Returns an executable from a debuginfod source. Writes are buffered into the writer.
+    pub fn get_executable(&self, build_id: &[u8], dest: &mut impl Write) -> Result<()> {
+        // /buildid/<BUILDID>/executable
+        let url = self
+            .base_url
+            .join("buildid/")?
+            .join(DebugInfodClient::base_path_for(build_id)?.as_str())?
+            .join("executable")?;
+        let mut res = self.client.execute(Request::new(Method::GET, url))?;
+        res.copy_to(dest)?;
+        Ok(dest.flush()?)
+    }
+
+    /// Returns a vector of default debuginfod URLs based on DEBUGINFOD_URLS environment variable.
+    pub fn get_default_servers() -> Result<Vec<Url>> {
+        if !env::var("DEBUGINFOD_URLS").is_ok() {
+            return Err(anyhow!("DEBUGINFOD_URLS is unset"));
+        }
+        let val = env::var("DEBUGINFOD_URLS");
+        let res: Vec<String> = val
+            .expect("DEBUGINFOD_URLS misconfigured")
+            .split([',', ' '])
+            .map(|v| v.to_string())
+            .collect();
+        let urls: Vec<Url> = res.iter().filter_map(|v| Url::parse(v).ok()).collect();
+        Ok(urls)
+    }
+
+    /// Returns a set of default clients based on the get_default_servers method.
+    pub fn get_default_clients() -> Result<Vec<DebugInfodClient>> {
+        let default_servers = DebugInfodClient::get_default_servers()?;
+        Ok(default_servers
+            .into_iter()
+            .map(|base_url| DebugInfodClient::new(Client::new(), base_url))
+            .collect::<Vec<DebugInfodClient>>())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    #[test]
+    fn build_id() {
+        let bytes = b"\x47\x91\x72\x34\xfb\xbd\x1c\x72\x88\x68\x4d\x92\x32\xfe\x2a\x27\x9e\x96\xfa";
+        let slice: &[u8] = bytes;
+        let expected = "47917234fbfbd1c7288684d9232fe2a279e96fa4";
+        let formatted_build_id = DebugInfodClient::format_build_id(slice).unwrap();
+        assert_eq!(formatted_build_id, expected);
+    }
+    #[test]
+    fn valid_default_servers_ws() {
+        env::set_var(
+            "DEBUGINFOD_URLS",
+            "https://debug.infod https://de.bug.info.d",
+        );
+        let resolvers = DebugInfodClient::get_default_servers().unwrap();
+        assert_eq!(resolvers.len(), 2);
+        assert_eq!(
+            resolvers[0],
+            Url::parse("https://debug.infod").ok().unwrap()
+        );
+        assert_eq!(
+            resolvers[1],
+            Url::parse("https://de.bug.info.d").ok().unwrap()
+        );
+    }
+    #[test]
+    fn valid_default_servers_comma() {
+        env::set_var(
+            "DEBUGINFOD_URLS",
+            "https://debug.infod,https://de.bug.info.d",
+        );
```

Comment:
Let's find a way to make this work without touching global environment variables. Cargo runs tests in parallel and here we are setting global state in one test and also in another. This can result in flakiness down the line as it's racy.

Can we introduce an internal constructor or something that omits this need?